tikvrpc/interceptor.go

```diff
@@ -0,0 +1,187 @@
+//
+// We can use InterceptorChain like this:
+// ```
+// func Interceptor1(next InterceptorFunc) InterceptorFunc {
+//     return func(target string, req *Request) (*Response, error) {
+//         fmt.Println("begin-interceptor-1")
+//         defer fmt.Println("end-interceptor-1")
+//         return next(target, req)
+//     }
+// }
+// func Interceptor2(next InterceptorFunc) InterceptorFunc {
+//     return func(target string, req *Request) (*Response, error) {
+//         fmt.Println("begin-interceptor-2")
+//         defer fmt.Println("end-interceptor-2")
+//         return next(target, req)
+//     }
+// }
+// txn.SetInterceptor(NewInterceptorChain().Link(Interceptor1).Link(Interceptor2).Build())
+// ```
+//
+// Then every time an RPC request is initiated, the following text will be printed:
+// ```
+// begin-interceptor-1
+// begin-interceptor-2
+// /* do request & respond here */
+// end-interceptor-2
+// end-interceptor-1
+// ```
+type InterceptorChain struct {
+	chain []Interceptor
+}
+
+// NewInterceptorChain creates an empty InterceptorChain.
+func NewInterceptorChain() *InterceptorChain {
+	return &InterceptorChain{}
+}
+
+// Link is used to link the next Interceptor.
+// Multiple interceptors will be executed in the order of link time.
+func (c *InterceptorChain) Link(it Interceptor) *InterceptorChain {
+	c.chain = append(c.chain, it)
+	return c
+}
+
+// Build merges the previously linked interceptors into one.
+func (c *InterceptorChain) Build() Interceptor {
+	return func(next InterceptorFunc) InterceptorFunc {
+		for n := len(c.chain) - 1; n >= 0; n-- {
+			next = c.chain[n](next)
+		}
+		return next
+	}
+}
+
+type interceptorCtxKeyType struct{}
+
+var interceptorCtxKey = interceptorCtxKeyType{}
+
+// SetInterceptorIntoCtx is a helper function used to bind Interceptor into ctx.
+// Different from the behavior of calling context.WithValue() directly, calling
+// SetInterceptorIntoCtx multiple times will not bind multiple Interceptors, but
+// will replace the original value each time.
+// Be careful not to forget to use the returned ctx.
+func SetInterceptorIntoCtx(ctx context.Context, interceptor Interceptor) context.Context {
+	if v := ctx.Value(interceptorCtxKey); v != nil {
+		v.(*atomic.Value).Store(interceptor)
+		return ctx
+	}
+	v := new(atomic.Value)
+	v.Store(interceptor)
+	return context.WithValue(ctx, interceptorCtxKey, v)
+}
+
+// GetInterceptorFromCtx gets the Interceptor bound by the previous call to SetInterceptorIntoCtx,
+// and returns nil if there is none.
+func GetInterceptorFromCtx(ctx context.Context) Interceptor {
+	if v := ctx.Value(interceptorCtxKey); v != nil {
+		v := v.(*atomic.Value).Load()
+		if interceptor, ok := v.(Interceptor); ok && interceptor != nil {
+			return interceptor
+		}
+	}
+	return nil
+}
+
+/* Suite for testing */
+
+// MockInterceptorManager can be used to create Interceptor and record the
+// number of executions of the created Interceptor.
+type MockInterceptorManager struct {
```

Comment:
Maybe you can write this `MockInterceptorManager` in `interceptor_test.go`.

Reply:
This suite is also used in integration tests, so it needs to be placed here.